In [62]:
import pandas as pd
data = pd.read_csv('https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv', 
                   names=['surgery','Age','Hospital Number','rectal temperature','pulse','6','temperature of extremities','peripheral pulse','mucous membranes','10','pain',' peristalsis','abdominal distension','14','15','16','17','18','packed cell volume','total protein','21','22','outcome','24','25','26','27','cp_data'],
                   na_values='?')

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     299 non-null    float64
 1   Age                         300 non-null    int64  
 2   Hospital Number             300 non-null    int64  
 3   rectal temperature          240 non-null    float64
 4   pulse                       276 non-null    float64
 5   6                           242 non-null    float64
 6   temperature of extremities  244 non-null    float64
 7   peripheral pulse            231 non-null    float64
 8   mucous membranes            253 non-null    float64
 9   10                          268 non-null    float64
 10  pain                        245 non-null    float64
 11   peristalsis                256 non-null    float64
 12  abdominal distension        244 non-null    float64
 13  14                          196 non

In [4]:
data.drop(['6','10','14','15','16','17','18','21','22','24','25','26','27'], axis=1, inplace=True)

In [5]:
data.describe()

surgery         Age  Hospital Number  rectal temperature  \
count  299.000000  300.000000     3.000000e+02          240.000000   
mean     1.397993    1.640000     1.085889e+06           38.167917   
std      0.490305    2.173972     1.529801e+06            0.732289   
min      1.000000    1.000000     5.184760e+05           35.400000   
25%      1.000000    1.000000     5.289040e+05           37.800000   
50%      1.000000    1.000000     5.303055e+05           38.200000   
75%      2.000000    1.000000     5.347275e+05           38.500000   
max      2.000000    9.000000     5.305629e+06           40.800000   

            pulse  temperature of extremities  peripheral pulse  \
count  276.000000                  244.000000        231.000000   
mean    71.913043                    2.348361          2.017316   
std     28.630557                    1.045054          1.042428   
min     30.000000                    1.000000          1.000000   
25%     48.000000                    1.000000          1.000000   
50%     64.000000                    3.000000          2.000000   
75%     88.000000                    3.000000          3.000000   
max    184.000000                    4.000000          4.000000   

       mucous membranes       pain   peristalsis  abdominal distension  \
count        253.000000  245.00000    256.000000            244.000000   
mean           2.853755    2.95102      2.917969              2.266393   
std            1.620294    1.30794      0.976744              1.065131   
min            1.000000    1.00000      1.000000              1.000000   
25%            1.000000    2.00000      3.000000              1.000000   
50%            3.000000    3.00000      3.000000              2.000000   
75%            4.000000    4.00000      4.000000              3.000000   
max            6.000000    5.00000      4.000000              4.000000   

       packed cell volume  total protein     outcome     cp_data  
count          271.000000     267.000000  299.000000  300.000000  
mean            46.295203      24.456929    1.551839    1.670000  
std             10.419335      27.475009    0.737187    0.470998  
min             23.000000       3.300000    1.000000    1.000000  
25%             38.000000       6.500000    1.000000    1.000000  
50%             45.000000       7.500000    1.000000    2.000000  
75%             52.000000      57.000000    2.000000    2.000000  
max             75.000000      89.000000    3.000000    2.000000

1)rectal temperature - 75% of all animals had elevated temp (25% quantile = 37.8-normal temp)
2)Std of pulse and Gap between 75% quant. and max are huge so there can be some outliers in data
3)most animals  had  normal packed cell volume (mean =46, median=45, 25% =23, 75%q = 52, normal 30-50)
4)50% of all animals had  elevated total protein (normal 6-7,5; median=7.5)

In [21]:
# Задание 2. Ищем  выбросы
data.groupby('total protein')['total protein'].count()

total protein
3.3     1
4.0     1
4.5     2
4.6     1
4.7     1
       ..
81.0    2
82.0    1
85.0    1
86.0    1
89.0    1
Name: total protein, Length: 81, dtype: int64

In [24]:
q1 = data['total protein'].quantile(0.25)
q3 = data['total protein'].quantile(0.75)
iqr = q3 - q1
higher_bound = q3 + (1.5*iqr)
print(higher_bound)
# max=89
# Нет выбросов, которыми стоило бы пренебречь, т.к значения плавно возрастают, к тому же рассчет через iqr говорит, что это не выброс

132.75


In [30]:
q1_pulse = data['pulse'].quantile(0.25)
q3_pulse = data['pulse'].quantile(0.75)
iqr_pulse = q3_pulse - q1_pulse
higher_bound_pulse = q3_pulse + (1.5*iqr_pulse)
print('higher_bound= ',higher_bound_pulse)
data.groupby('pulse')['pulse'].count()
# max=184
# значения пульса тоже не выглядят как ошибочные. Вопрос только с  одним наблюдением, где пульс равен max  значению. Тут лучше посоветоваться со специалистом

higher_bound=  148.0


pulse
30.0      2
36.0      3
38.0      1
40.0     18
42.0     10
44.0     11
45.0      2
46.0      1
48.0     28
49.0      1
50.0      6
52.0     12
54.0      7
56.0      7
60.0     25
64.0      8
65.0      1
66.0      6
68.0      5
70.0      4
72.0     11
75.0      1
76.0      4
78.0      3
80.0      8
82.0      1
84.0      8
86.0      3
88.0     12
90.0      3
92.0      4
96.0      8
98.0      1
100.0    11
104.0     5
108.0     3
110.0     1
112.0     3
114.0     2
120.0    10
124.0     2
128.0     1
129.0     1
130.0     2
132.0     1
136.0     1
140.0     2
146.0     1
150.0     2
160.0     1
164.0     1
184.0     1
Name: pulse, dtype: int64

In [31]:
# Задание 3. Пропуски
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 15 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     299 non-null    float64
 1   Age                         300 non-null    int64  
 2   Hospital Number             300 non-null    int64  
 3   rectal temperature          240 non-null    float64
 4   pulse                       276 non-null    float64
 5   temperature of extremities  244 non-null    float64
 6   peripheral pulse            231 non-null    float64
 7   mucous membranes            253 non-null    float64
 8   pain                        245 non-null    float64
 9    peristalsis                256 non-null    float64
 10  abdominal distension        244 non-null    float64
 11  packed cell volume          271 non-null    float64
 12  total protein               267 non-null    float64
 13  outcome                     299 non

In [45]:
# в условии сказано, что temperature of extremities и rectal temperature несколько коррелируют, хотя по цифрам этого не скажешь
# поэтому заменяем просто средним
data.groupby('temperature of extremities')['rectal temperature'].median()

temperature of extremities
1.0    38.15
2.0    38.20
3.0    38.10
4.0    38.35
Name: rectal temperature, dtype: float64

In [49]:
ch_data = data
ch_data['rectal temperature'].fillna(data['rectal temperature'].mean(), inplace=True)

In [61]:
# можно заполнить pain, используя surgery, т.к они взаимосвязаны
ch_data['pain'].fillna(data.groupby('surgery')['pain'].transform('mean'), inplace=True)
ch_data['pain']=ch_data['pain'].round(0)

data.groupby('surgery')['pain'].mean()

surgery
1.0    3.233333
2.0    2.386555
Name: pain, dtype: float64

In [ ]:
# можно и дальше продолжать заполнять пропуски, но смысла особо нет